### PR TITLE
Fix sklearn autologging input example & signature mutation for models containing transformers

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Fix sklearn autologging input example & signature mutation for models containing transformers

## How is this patch tested?

Added integration tests for sklearn autologging with transformers + synthetic mutation test cases for xgboost and lightgbm.

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

Fix sklearn autologging input example & signature mutation for models containing transformers

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/sklearn/__init__.py [hunk 1]
```diff
@@ -1262,15 +1263,25 @@ def fit_mlflow_xgboost_and_lightgbm(original, self, *args, **kwargs):
...
+        # for subsequent use during model logging, ensuring that the input example and inferred
+        # model signature to not include any mutations from model training
+        input_example = _get_X_y_and_sample_weight(self.fit, args, kwargs)[0][
+            :INPUT_EXAMPLE_SAMPLE_ROWS
+        ]
         # parameter, metric, and non-model artifact logging are done in
         # `train()` in `mlflow.xgboost.autolog()` and `mlflow.lightgbm.autolog()`
         fit_output = original(self, *args, **kwargs)
         # log models after training
-        X = _get_X_y_and_sample_weight(self.fit, args, kwargs)[0]
```
Comment (dbczumar) [inline mlflow/sklearn/__init__.py:1269]: By this time, `X` has already been mutated by any sckit-learn transformations defined by the estimator.

File: tests/sklearn/test_sklearn_autolog.py [hunk 2]
```diff
@@ -906,6 +906,54 @@ def test_autolog_logs_signature_and_input_example(data_type):
...
+    )
+    model_signature_input_names = [inp.name for inp in model_conf.signature.inputs.inputs]
+    assert "XLarge Bags" in model_signature_input_names
+    assert "XLarge Bags" in input_example.columns
+    assert "XXLarge Bags" not in model_signature_input_names
+    assert "XXLarge Bags" not in input_example.columns
+
+    metrics = get_run_data(mlflow.last_active_run().info.run_id)[1]
+    assert "training_r2_score" in metrics
+    assert "training_rmse" in metrics
```
Comment (dbczumar) [inline tests/sklearn/test_sklearn_autolog.py:954]: We test metrics because mutation with transformations can affect metric computations (i.e. failure due to unexpected columns, invalid data distribution)

**Review by jinzhang21 — APPROVED** [review]

**Comment by mahesh-venkatachalam** [comment]
Thanks for the super quick turnaround here, @dbczumar!

File: tests/xgboost/test_xgboost_autolog.py [hunk 3]
```diff
@@ -194,6 +194,57 @@ def test_xgb_autolog_sklearn():
...
+            X["TESTCOL"] = 5
+            return original_xgb_regressor_fit(self, *args, **kwargs)
+
+        def patched_xgb_regressor_predict(self, *args, **kwargs):
+            X = args[0]
+            X["TESTCOL"] = 5
+            return original_xgb_regressor_predict(self, *args, **kwargs)
+
+        xgb.XGBRegressor.fit = patched_xgb_regressor_fit
+        xgb.XGBRegressor.predict = patched_xgb_regressor_predict
```
Comment (harupy) [inline tests/xgboost/test_xgboost_autolog.py:214]:
```suggestion
    with mock.patch("xgboost.XGBRegressor.fit", patched_xgb_regressor_fit), mock.patch(
        "xgboost.XGBRegressor.predict", patched_xgb_regressor_predict
    ):
```

Can we use `mock.patch` and remove try-finally?

File: tests/sklearn/test_sklearn_autolog.py [hunk 4]
```diff
@@ -906,6 +906,54 @@ def test_autolog_logs_signature_and_input_example(data_type):
...
+
+    X_train = pd.DataFrame.from_dict(
+        {
+            "Total Volume": {0: 64236.62, 1: 54876.98, 2: 118220.22},
+            "Total Bags": {0: 8696.87, 1: 9505.56, 2: 8145.35},
+            "Small Bags": {0: 8603.62, 1: 9408.07, 2: 8042.21},
+            "Large Bags": {0: 93.25, 1: 97.49, 2: 103.14},
+            "XLarge Bags": {0: 0.0, 1: 0.0, 2: 0.0},
+        }
+    )
```
Comment (harupy) [inline tests/sklearn/test_sklearn_autolog.py:920]:
```suggestion
    X_train = pd.DataFrame(
        {
            "Total Volume": [64236.62, 54876.98, 118220.22],
            "Total Bags": [8696.87, 9505.56, 8145.35],
            "Small Bags": [8603.62, 9408.07, 8042.21],
            "Large Bags": [93.25, 97.49, 103.14],
            "XLarge Bags": [0.0, 0.0, 0.0],
        }
    )
```

nit to simplify the code

File: tests/sklearn/test_sklearn_autolog.py [hunk 5]
```diff
@@ -906,6 +906,54 @@ def test_autolog_logs_signature_and_input_example(data_type):
...
+    y_train = pd.Series({0: 1.33, 1: 1.35, 2: 0.93})
```
Comment (harupy) [inline tests/sklearn/test_sklearn_autolog.py:921]:
```suggestion
    y_train = pd.Series([1.33, 1.35, 0.93])
```

same here https://github.com/mlflow/mlflow/pull/6230/files#r918496075

**Review by harupy — APPROVED** [review]
Left a few comments, otherwise LGTM!

File: tests/sklearn/test_sklearn_autolog.py (hunk 4 above)
Comment (dbczumar) [reply tests/sklearn/test_sklearn_autolog.py]: Simplified! Thanks :)

File: tests/sklearn/test_sklearn_autolog.py (hunk 5 above)
Comment (dbczumar) [reply tests/sklearn/test_sklearn_autolog.py]: Done! :)

File: tests/xgboost/test_xgboost_autolog.py (hunk 3 above)
Comment (dbczumar) [reply tests/xgboost/test_xgboost_autolog.py:215]: Good catch! Done!

**Review by dbczumar — COMMENTED** [review]
@harupy Thank you for your comments! I've addressed them :)

**Review by WeichenXu123 — APPROVED** [review]
LGTM.
